wazuh/wazuh
# Fix mapping issues related to Filebeat making Kibana monitoring work properly

|Related issue|
|---|
| https://github.com/wazuh/wazuh/issues/3723 |

- Branch: https://github.com/wazuh/wazuh/tree/3.9.4-elastic-fixes
- Module: https://packages-dev.wazuh.com/3.x/filebeat/wazuh-filebeat-0.1.tar.gz
- Filebeat config: https://raw.githubusercontent.com/wazuh/wazuh/3.9.4-elastic-fixes/extensions/filebeat/7.x/filebeat.yml
- Elasticsearch template: https://raw.githubusercontent.com/wazuh/wazuh/3.9.4-elastic-fixes/extensions/elasticsearch/7.x/wazuh-template.json

## Description

This pull request aims to avoid errors related to Kibana > Monitoring app due to a mapping issue. Details about these errors are described in the related ticket (https://github.com/wazuh/wazuh/issues/3723).

## Changes

- https://github.com/wazuh/wazuh/commit/6cca7ef8024c71c245840dc344c633dc38a6a905 adds `event.dataset` as a new property of our template, this way Kibana thinks all is fine with the mapping because it's the field with throws the error. On the other hand (https://github.com/wazuh/wazuh/commit/156418991624565c23eb7ce42462c667a20dfd7d), I've added `filebeat*` in our patterns list so Kibana thinks a template exists for `filebeat*` indices.
- https://github.com/wazuh/wazuh/commit/5ead70335ba35ec76a444880a26d1e73629a4537 this prevents the empty index creation of Filebeat. After reading the docs, Filebeat has its own ILM (index lifecycle management) policies. This disables auto-creation of an empty index with the prefix `filebeat*`, but it's not the only way Filebeat may create an index, but it's explained in next step.
- https://github.com/wazuh/wazuh/commit/f2cfd6adcd79913b0d4e42c0e841a165fb51783d here I'm changing the `on_failure` section of our ingest pipeline. It's a fact that sending an alert that fails at ingestion time, Filebeat tries to send it to its own indices, here where are discarding those events, so If an event is sent by our module and it is not a valid Wazuh alert for any reason, then it won't try to send it to `filebeat*` index. This doesn't occur always, but with this change, we are covered.
- ~https://github.com/wazuh/wazuh/commit/770b4c8955933ffb32ca08418c8564a8c30ea2c4 bump version for the Filebeat module, now it's 0.2.~ Did a revert, so it's still 0.1

## Screenshots

Here is a screenshot of Kibana > Monitoring working properly, also I've configured filebeat adding `monitoring.enabled: true` for testing purposes:

![image](https://user-images.githubusercontent.com/8860227/61737137-e83cce00-ad87-11e9-894d-57008d986648.png)
![image](https://user-images.githubusercontent.com/8860227/61737219-073b6000-ad88-11e9-9cfa-6cc5990030d8.png)

Now, the same but with no Filebeat monitoring (our default):

![image](https://user-images.githubusercontent.com/8860227/61737279-2934e280-ad88-11e9-8705-b8f366c27804.png)

On the other hand, here is a screenshot from Discover showing Wazuh alerts:

![image](https://user-images.githubusercontent.com/8860227/61737379-55e8fa00-ad88-11e9-81ea-3e8d0a5bf6a7.png)


## Tests

Some of the tests I did were:
- Append a wrong alert to the alerts.json which would cause an error on our ingesting pipeline, before the changes it tried to send it to `filebeat*`, now it doesn't.
- Force the existence of a `filebeat*` index creating it by hand, just to be sure our logic is working, before the changes Monitoring fails, now it doesn't.
- Run Filebeat in debug mode several times, cleaning the environment and relaunching it, now it doesn't create any `filebeat*` indices and the Monitoring tab works properly.
- Activate the Filebeat monitoring features, it works properly.
- Send an altered alert (older timestamp), it creates the right index for the proper day.
- Send an alert expected to generate GeoLocation.location, it does.
- Remove the template while Filebeat is working, since the index exists it keeps sending the data.
- With the template removed, also remove today's index. Now Filebeat sends the template before sending the events because Elasticsearch is going to create a new index, so we are fully covered for these situations.

## Conclusion

That's said, these changes are expected to cover Wazuh integration and to make Kibana monitoring work properly.
